Fix havocking of dependent assigns clause targets

remi-delmas-3000 opened the pull request:
The first commit fixes a minor issue with history variables (allow history variables for constant and typecast expressions).

The second commit fixes missing declarations for return value variables introduced during replacement of function calls by contracts.

The third commit fixes a problem with assigns clause havocking when replacing calls by contracts.

When replacing a function call by its contract, we would directly havoc the targets of the assigns clause in an unspecified order. If the assigns clause contained dependent targets such as `assigns(s->ptr, *(s->ptr))`, havocking `s->ptr` before `*s->ptr` would lead to errors and incorrect results. 

We now introduce independent snapshot variables storing the address of each target and havoc them in a second time to avoid this dependency on the ordering of havoc instructions.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

tautschnig commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -562,9 +562,10 @@ void code_contractst::replace_history_parameter(
 
+    const auto &id = parameter.id();
     if(
-      parameter.id() == ID_dereference || parameter.id() == ID_member ||
-      parameter.id() == ID_symbol || parameter.id() == ID_ptrmember)
+      id == ID_dereference || id == ID_member || id == ID_symbol ||
+      id == ID_ptrmember || id == ID_constant || id == ID_typecast)
     {
```
Is there a test that demonstrates the need for this?

remi-delmas-3000 replied:
The tests added in this PR will fail without this.

remi-delmas-3000 replied:
added specific tests

tautschnig commented on regression/contracts/assigns_replace_08/test.desc:
```diff
@@ -3,7 +3,7 @@ main.c
 ^EXIT=10$
 ^SIGNAL=0$
-\[foo.\d+\] line \d+ Check that \*z is assignable: FAILURE$
+\[foo.\d+\] line \d+ Check that \*z assigned by replaced bar is assignable: FAILURE$
```
I'm struggling to parse this - should it perhaps say "replaced *function* bar?"

feliperodri replied:
How about `Check that \*z is assignable (function bar replacement): `?

remi-delmas-3000 replied:
yes

tautschnig commented on src/goto-instrument/contracts/havoc_assigns_clause_targets.h:
```diff
@@ -0,0 +1,62 @@
+
+#include <goto-programs/goto_program.h>
+#include <util/namespace.h>
+#include <util/symbol_table.h>
```
Forward declarations (`class namespacet;`, `class symbol_tablet;`) will do the trick here. (This is to reduce compilation times where possible.)

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/havoc_assigns_clause_targets.h:
```diff
@@ -0,0 +1,62 @@
+#ifndef CPROVER_GOTO_INSTRUMENT_CONTRACTS_HAVOC_ASSIGNS_CLAUSE_TARGETS_H
+#define CPROVER_GOTO_INSTRUMENT_CONTRACTS_HAVOC_ASSIGNS_CLAUSE_TARGETS_H
+
+#include <set>
```
This isn't used here.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/havoc_assigns_clause_targets.h:
```diff
@@ -0,0 +1,62 @@
+
+#include <set>
+
+#include <goto-programs/goto_program.h>
```
I think you'll only need `#include <util/expr.h>`, everything else can be forward-declared.

remi-delmas-3000 replied:
indeed this works

tautschnig commented on src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp:
```diff
@@ -0,0 +1,239 @@
+{
+  if(target.id() == ID_pointer_object)
+  {
+    const auto &ptr = target.operands().front();
```
Please use `to_unary_expr(target).op()` (for a minimal amount of sanity checking).

tautschnig commented on src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp:
```diff
@@ -0,0 +1,239 @@
+  {
+    const auto &ptr = target.operands().front();
+    // bring the offset back to zero
+    auto res = minus_exprt{ptr, pointer_offset(ptr)};
+    res.type() = ptr.type();
```
Oh, we're really lacking a proper constructor for `minus_exprt` here. Adding this to my todo list.

tautschnig commented on src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp:
```diff
@@ -0,0 +1,239 @@
+  else if(is_assignable(target))
+  {
+    // check that the type of the target has a definite size
+    const auto &size = size_of_expr(target.type(), ns);
+
+    INVARIANT(
+      size.has_value(),
```
Please move this into the `INVARIANT` to avoid a situation where `size` ends up as an unused value (when completely disabling invariants). That is: `INVARIANT(size_of_expr(target.type(), ns).has_value(), ...)`

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp:
```diff
@@ -0,0 +1,239 @@
+      "`sizeof` must always be computable on assignable assigns clause "
+      "targets.");
+
+    return address_of_exprt{target, pointer_type(target.type())};
```
Wouldn't `is_assignable` be true for something like `a[5]`, but taking the address thereof would not actually be the "base address" (or at least not how I'd understand that term).

remi-delmas-3000 replied:
This computes the start address of the thing that gets assigned. If contract assigns `a[5]` the base address of the thing that gets assigned is `&(a[5])`. If the contract says it assigns `__CPROVER_pointer_object(a[5])` then the base address will be the address pointed to by `a-5`.

tautschnig replied:
I'd then find something along the lines of "build_address_of" a more natural function name, but this might be a matter of personal preference.

kroening replied:
As a side, note that `address_of_exprt` has a convenience constructor that takes just the object, i.e., the above is equivalent to `address_of_exprt{target}`. The constructor will create the pointer type from the type of `target`.

remi-delmas-3000 replied:
renamed to `build_address_of`

tautschnig commented on src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp:
```diff
@@ -0,0 +1,239 @@
+  // context parameters
+  const namespacet &ns)
+{
+  dest.add(goto_programt::make_decl(target_valid_var));
+
+  dest.add(goto_programt::make_decl(target_snapshot_var));
+
+  dest.add(goto_programt::make_assignment(
+    target_valid_var, all_dereferences_are_valid(target_pointer, ns)));
+
+  dest.add(goto_programt::make_assignment(
+    target_snapshot_var,
+    null_pointer_exprt{to_pointer_type(target_pointer.type())}));
```
For debugging purposes it would be helpful if all these had source locations that relate them in some way to the location a snapshot is taken at. Hmm, you're doing this after the fact. That's a bit confusing, in my opinion.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp:
```diff
@@ -0,0 +1,239 @@
+    target_snapshot_var,
+    null_pointer_exprt{to_pointer_type(target_pointer.type())}));
+
+  goto_programt skip_program;
+  const auto skip_target = skip_program.add(goto_programt::make_skip());
+
+  dest.add(goto_programt::make_goto(skip_target, not_exprt{target_valid_var}));
+
+  dest.add(goto_programt::make_assignment(target_snapshot_var, target_pointer));
+
+  dest.destructive_append(skip_program);
```
You could use `insert_before_swap` to avoid the `skip_program`. But then it probably doesn't make a whole lot of a difference.

remi-delmas-3000 replied:
it seems less natural, since after inserting the skip statement in dest, you would have to use a mix of insert_before_swap and skip_target++ or to insert instructions in reverse, or to create another program and insert_before_swap and then ++ the skip_target

tautschnig replied:
Ok!

tautschnig commented on src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp:
```diff
@@ -0,0 +1,239 @@
+  if(target.id() == ID_pointer_object)
+  {
+    // OTHER __CPROVER_havoc_object(target_snapshot_var)
+    codet code(ID_havoc_object);
+    code.add_to_operands(target_snapshot_var);
```
`codet code(ID_havoc_object, {target_snapshot_var});` would avoid the two-step construction.

remi-delmas-3000 replied:
fixed

feliperodri commented on regression/contracts/assigns_replace_havoc_dependent_targets/vect.h:
```diff
@@ -0,0 +1,48 @@
+  // clang-format off
+__CPROVER_requires(
+  __CPROVER_is_fresh(v, sizeof(vect)) &&
+  0 < v->size && v->size <= __CPROVER_max_malloc_size &&
```
Why do you need to check whether `v->size` is smaller than `__CPROVER_max_malloc_size`? Haven't this problem been solved (see https://github.com/diffblue/cbmc/issues/6130)?

remi-delmas-3000 replied:
Whatever the behaviour of CBMC is when this bound is exceeded, I want to avoid it in this example.

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -786,15 +787,17 @@ bool code_contractst::apply_function_contract(
-    havoc_assigns_targetst havoc_gen(assigns, ns);
-    havoc_gen.append_full_havoc_code(location, havoc_instructions);
+    // Havoc all targets in the assigns clause
+    // TODO handle local statics possibly touched by this function
```
Is there a issue to track this?

remi-delmas-3000 replied:
yes

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -786,15 +787,17 @@ bool code_contractst::apply_function_contract(
-
-    havoc_assigns_targetst havoc_gen(assigns, ns);
-    havoc_gen.append_full_havoc_code(location, havoc_instructions);
+    // Havoc all targets in the assigns clause
+    // TODO handle local statics possibly touched by this function
```
nit:
```suggestion
    // Havoc all targets in the assigns clause.
    // TODO: handle local statics possibly touched by this function.
```

feliperodri commented on regression/contracts/assigns_replace_havoc_dependent_targets_fail/enforce.desc:
```diff
@@ -0,0 +1,9 @@
+^SIGNAL=0$
+--
+--
+Verifies the contract being replaced in `replace_resize_vec.desc`.
```
Is that the correct file name? I didn't find it.

remi-delmas-3000 replied:
fixed

feliperodri commented on regression/contracts/assigns_replace_havoc_dependent_targets_fail/vect.h:
```diff
@@ -0,0 +1,48 @@
+  // clang-format off
+__CPROVER_requires(
+  __CPROVER_is_fresh(v, sizeof(vect)) &&
+  0 < v->size && v->size <= __CPROVER_max_malloc_size &&
```
See https://github.com/diffblue/cbmc/issues/6130.

Hasn't this problem been fixed? Why do you need `v->size <= __CPROVER_max_malloc_size` here?

remi-delmas-3000 replied:
Whatever the behaviour of CBMC is when this bound is exceeded, I want to avoid it in this example.

feliperodri commented on src/goto-instrument/contracts/havoc_assigns_clause_targets.h:
````diff
@@ -0,0 +1,62 @@
+/// ```
+/// typedef struct vect { int *arr; int size; } vect;
+/// void resize(vect *v)
+/// __CPOVER_assigns(v->arr, v->size, __CPROVER_POINTER_OBJECT(v->arr))
````
typo: `__CPROVER_assigns`

remi-delmas-3000 replied:
fixed

feliperodri reviewed (approved):
Only minor comments.

Commit:
Function contracts: allow history variables for typecast and constant expressions.

Commit:
Function contracts: add missing DECL/DEAD instructions for `ignored_return_value` variables introduced by contract replacement (these would previously appear as global variables).

Commit:
Function contracts: new method for havocking assigns clause targets
that works when there are dependencies between targets.

We now take a snapshot of the target addresses if they are valid and
havoc them in a second step if they are valid.  Previously we were
havocking targets directly in an arbitrary sequence, and havocking
s->ptr before *(s->ptr) before __CPROVER_POINTER_OBJECT(s->ptr) would
cause spurious errors.

tautschnig commented on src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp:
```diff
@@ -0,0 +1,252 @@
+  {
+    const auto &ptr = target.operands().front();
+    // bring the offset back to zero
+    auto res = minus_exprt{ptr, pointer_offset(ptr)};
+    res.type() = ptr.type();
+    return std::move(res);
```
Scratch my earlier comment, the following will actually do the right thing: `return minus_exprt{ptr, pointer_offset(ptr)};` (the type of the `minus_exprt` will be `ptr.type()` when constructed in this way).

remi-delmas-3000 replied:
thanks

remi-delmas-3000 commented:
I took into account all pending comments, it should be good to go now.

tautschnig commented:
Requires approval from @peterschrammel @martin-cs @chrisr-diffblue as a new file was added to `goto-instrument/Makefile`.

feliperodri commented on regression/contracts/history-constant/main.c:
```diff
@@ -0,0 +1,15 @@
+#include <stdio.h>
```
Why do we need this include?

feliperodri commented on regression/contracts/history-typecast/main.c:
```diff
@@ -0,0 +1,23 @@
+#include <stdio.h>
```
Why do we need this include?

feliperodri reviewed (approved):
It needs code owner approval from @peterschrammel @chrisr-diffblue or @martin-cs.

chrisr-diffblue reviewed (approved):
Approving the Makefile changes. Nothing to add beyond the other detailed reviews.
